src/kernels/attention_kernels.h

```diff
@@ -596,4 +648,135 @@ void crossAttnShardedHead(T1 *output, const T1 *query, const float *attnMask, in
     } // end for b
 }
 
+/**
+ * @brief Cross attention with head granularity (including copy key/value to KV Cache)
+ * @note if causal = True, attnMask is not used
+ * @tparam T Data type
+ * @tparam KVCacheT KV cache data type
+ * @tparam Lambda1 Lambda to get head of cached keys, return tuple of (addr, stride, scale)
+ * @tparam Lambda2 Lambda to get head of cached values
+ * @param output Output tensor
+ * @param query Query tensor, in shape of (input_seqlen, head_num, head_size)
+ * @param key Key tensor (not include cached keys)
+ * @param value Value tensor (not include cached values)
+ * @param qHeadNum Query head number
+ * @param kvHeadNum KV head number
+ * @param headSize Head size
+ * @param oStride Output stride
+ * @param qStride Query stride
+ * @param kvStride KV stride
+ * @param batchSize Batch size
+ * @param inputSeqLens Input sequence lengths
+ * @param pastSeqLens Past sequence lengths
+ * @param causal Whether causal attention
+ * @param alibiSlopes Alibi slopes
+ * @param attnMask Attention mask
+ * @param scale Scale factor for softmax
+ * @param threadNum Thread number
+ * @param getKHead Lambda to get head of cached keys
+ * @param getVHead Lambda to get head of cached values
+ */
+template <typename T, typename KVCacheT, typename Lambda1, typename Lambda2>
+void crossAttnByHead(T *output, const T *query, const T *key, const T *value, int qHeadNum, int kvHeadNum, int headSize,
+        int oStride, int qStride, int kvStride, int batchSize, const int *inputSeqLens, const int *pastSeqLens,
+        bool causal, const float **alibiSlopes, const float **attnMask, const float scale, int threadNum,
```

Comment:
"const float *" is ok both for alibiSlopes and attnMask

Reply:
Note this is for several samples/sequences, I am not sure if a single pointer is OK.

Reply:
The alibiSlopes are a set of values that are the same for all sequences. And the AttnMask can also be seen as a large buffer in the bc mode.

Reply:
Done and removed attnMask.